### ISSUE TITLE
Feature: Independent horizontal/vertical line width (elliptical/anisotropic pen, asymmetrical shape)

### ISSUE THREAD

**Issue by nyanpasu64** [matplotlib/matplotlib#12980]
### Bug report

**Bug summary**

I would like the ability to render plots where lines are asymmetrical, such that vertical lines have a narrow width, while horizontal lines have a larger height.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt
plt.plot(np.random.rand(2000))
plt.show()
```

**Actual outcome**

I lose most horizontal detail, and the overlapping lines essentially fill the entire screen with blue. This is because horizontal points are uniformly spaced, but vertical points are random.

**Expected outcome**

The ability to make vertical lines much thinner, so I can see fine detail more clearly, while low-frequency sections produce tall horizontal lines.

I'm encountering this issue in my audio oscilloscope program, which smears out horizontal detail when plotting complex waves. I usually plot 500-2000 samples in 600-1200 pixels, but decreasing the line width makes the line hard to see (especially on mobile and Youtube thumbnails).

**Matplotlib version**

  * Operating system: Ubuntu 18.04
  * Matplotlib version: 3.0.0 Anaconda main channel
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://backend_interagg
  * Python version: 3.6.6

**Ideas**

agg `math_stroke` assumes a fixed aspect ratio, and is used to calculate corners (and probably ends) which controls stroke shape.

The easiest solution is to edit `_backend_agg.h` and change `RendererAgg::_draw_path` to stretch `path` horizontally before casting to `stroke_t stroke`, and unstretch `stroke` afterwards. I can't implement this using a pure Python subclass, and don't want to build my own matplotlib C++ to bundle with my app.

**Complaints**

agg's duck-typed templates are baffling to understand. Maybe I'm just not well-versed in Qt Creator. How do I show the computed template-type of various objects, and the types accepted?

**Comment by jimbo1qaz** [comment]
Are there any updates on this feature request?

**Comment by dstansby** [comment]
Does doing `plt.plot(..., linewidth=0.5)` solve the issue?

**Comment by jimbo1qaz** [comment]
Decreasing line width is not that I want. It causes less complex line sections to be very hard to see. (I could tell users to decrease line width based on the average complexity of different channels, but that's an inelegant workaround, and I think inconsistent line widths may be ugly).

Another oscilloscope program (Ovgen, I believe it uses .net graphics) allows switching between a slightly narrow pen, and a near-zero-width pen.

**Comment by jklymak** [comment]
I’m 99.9% sure that example is not switching widths.  The horizontal portions have noise around a constant value  and hence the line has thickness, wherase the periodic jumps appear thinner.

**Comment by jimbo1qaz** [comment]
ovgen has a feature to use a narrow pen: https://github.com/Zeinok/OVGen/blob/49fec909c7d86628e7e3e212aa68df8251c4198f/OVGen/MainForm.vb#L225-L230

wavePen is a `System.Drawing.Pen`.

I'm not sure if that video uses this feature or not, but I'm certain this feature exists.

**Comment by jimbo1qaz** [comment]
As I stated in my original post...

agg `math_stroke` assumes a fixed aspect ratio, and is used to calculate corners (and probably ends) which controls stroke shape.

The easiest solution is to edit `_backend_agg.h` and change `RendererAgg::_draw_path` to stretch `path` horizontally before casting to `stroke_t stroke` (find outline of `path`), and unstretch `stroke` afterwards. I can't implement this using a pure Python subclass, and don't want to build my own matplotlib C++ to bundle with my app.

**Comment by anntzer** [comment]
I think the feature request is worth considering (as in, we may consider a PR implementing it, because I don't know any of the core devs who really looks forward to playing with the Agg code...), but needs to consider the following points:
- What would the API (both on the artist, and on the renderer) look like?
- This should hopefully work for both Line2D and LineCollection.
- This should work not only for the Agg backend, but also for the vector ones (pdf/ps/svg), and at least not completely break third-party renderers (even if they don't support the feature).

Moreover, I *think* it may be possible to implement this on Python's side by (ab)using the agg_filter mechanism.

It would also be nice to figure out what knobs we can expose to make this doable from a third-party package.

**Comment by jimbo1qaz** [comment]
Are you expecting me to write a proof of concept?

Note: Ovgen calls it "crt mode" because analog oscilloscopes get (dimmer? thinner?) on steep near/vertical lines.

# API

>What would the API (both on the artist, and on the renderer) look like?

## Line (is that what you mean by artist?)
- `ax.plot(data, color=line_color, linewidth=line_width, xsqueeze=3)`
- I only need `xsqueeze` but maybe add `ysqueeze` for symmetry. I think there is no valid reason to supply both parameters at once, you should adjust `linewidth` instead.
- Multiply the chosen dimension by the factor before outlining, then divide coordinates afterwards.
- **Naming issue:** the use of "squeeze" may conflict with `plt.subplots(squeeze=)` (https://matplotlib.org/api/_as_gen/matplotlib.pyplot.subplots.html).

The call stack is long and convoluted (most recent call first):

```
Line2D.__init__
_process_plot_var_args._makeline (why does a class have a snake_case name?)
_process_plot_var_args._plot_args
_process_plot_var_args._grab_next_args
Axes.plot
```

Probably add `xsqueeze/ysqueeze` kwargs into `Line2D.__init__()`? (I'm a fan of using `attr.dataclass(kw_only=True)` to group kwargs).

~~store in `gc`~~ bad idea, `gc` is a disposable local object created in each `draw()` call.

## Renderer

Is there a reason to add this feature to renderers other than Agg, and is that a prerequisite for merging a PR? Cairo was much slower than Agg for me, and I don't need this feature for SVG export/etc.

`Line2D.draw()` calls:
- `gc` is created and disposed in each `draw()` call, and is a `renderer.new_gc()` (GraphicsContext?)
- `gc.set_linewidth(self._linewidth)`
- `renderer.draw_path(gc, tpath, affine.frozen())`
  - python `RendererAgg.draw_path` is a thin wrapper (does something related to `agg.path.chunksize`) around c++ `RendererAgg::draw_path`.

## Renderer proposal

Performance is very important for me, as I'm rendering 60 fps at 1280x720 with 1-10k points. It would be a shame if I ended up not using this feature because of the speed hit... :(

I would use an `if(xsqueeze != 1)` to skip the multiplication speed hit, if no squeezing occurred.

I don't know how to avoid overhead when distortion is disabled. Unconditionally multiplying coordinates will add overhead if multiplying by 1. To avoid that, I feel a runtime branch must happen in either Python or C++, and `if/else` in C++ will be faster (unless when plotting `LineCollection`, Python performs 1 branch but C++ performs `O(nlines)` branches).

a. Add an if-else into Python.
a. Add an if-else into `draw_path` to call variants of `_draw_path` (`_draw_path_xsqueeze` and `_draw_path_ysqueeze`).
    - `draw_path` looks at `gc`, `_draw_path` either arguments or `gc`.
a. Add an if-else into `_draw_path`
    - Pass squeeze factors via `gc`.
    - I don't know if the if/else statement's resulting `stroke` variables are type-compatible.

Optimization: You transform path points from input space to graph space. Instead of transforming path points twice, you could multiply the "screen width" by `xsqueeze`.

**Comment by jimbo1qaz** [comment]
What's wrong with the Cairo backend? (for performance, I disable all artists other than background and lines. I only noticed the slow speed.) Would the new backend be faster than agg?

**Comment by jklymak** [comment]
I’m still against this as a general feature in the plot codepath.  I don’t think simulating an analog plotting artifact is of general enough interest to merit a new kwarg and added code complexity.

**Comment by tacaswell** [comment]
This feature is comparable to the inkscape calligraphy pen (https://inkscape.org/doc/tutorials/calligraphy/tutorial-calligraphy.html with the  angle set to 90 and the fixation to 100)?

I agree with most of what @anntzer said.  If we are going to implement pen shapes they should be supported across all of the 'core' backends where it can be supported (we try to keep the backends as close to parity as possible) or a very compelling case that this is impossible in the vector backends and across everything that get stroked (`Line2D`, `LineCollections`, `Polygons`, ...).  This is obviously a pretty big amount of work for something I do not see a clear path to.

I am very :-1: on a Agg specific path in `Line2D` to implement this. 

That said, we do have some Agg only features, such as `agg_filter` that @anntzer mentioned above.

```python
import numpy as np
import matplotlib.pyplot as plt
from skimage.transform import resize, rescale


def agg_filter(arr, dpi):
    out = stretch_and_resample(arr)
    return out, 0, 0

fig, ax = plt.subplots()
th = np.linspace(-20, 20, 1024)
ln, = ax.plot(th, np.sin(th*10) / th, lw=5)
ln.set_agg_filter(agg_filter)
```

but I have been failing at getting the resampling from skimage to work the way I want it to...

If you are concerned about performance you may be better served by https://github.com/pyqtgraph/pyqtgraph which either user the Qt native 2D artists or openGL.  That said, I am happy to know you are getting 60fps out of Matplotlib!

**Comment by jimbo1qaz** [comment]
Pyqtgraph's image-bytes extractor (I popped up to ffmpeg) is horrifically slow. After rewriting it, I managed to get 20% more fps than matplotlib agg (axes disabled for both plots). However, on pyqtgraph, setting line width > 1 (which enables outline-then-fill) caused performance to drop below matplotlib.

Enabling gpu/opengl on pyqtgraph seemed to have no effect on performance. (Maybe I could investigate further.)

Should I avoid discussing pyqtgraph on a matplotlib issue tracker?
